Kibana should allow a min_age setting of 0ms in ILM policy phases

jkelastic opened the pull request:
Fixes https://github.com/elastic/kibana/issues/50476

## Release note

The Index Lifecycle Policy editor now permits the minimum age to be 0.

elasticmachine commented:
Pinging @elastic/es-ui (Team:Elasticsearch UI)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16993/)
* Commit: 809b32f9b2a8887cb39140f0b4db9271ceecd585

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16993/","number":"16993","commit":"809b32f9b2a8887cb39140f0b4db9271ceecd585"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17051/)
* Commit: 231071a189a116afa963ca607110fdb161cbbddf

### History
* :broken_heart: [Build #16993](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16993/) failed 809b32f9b2a8887cb39140f0b4db9271ceecd585

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/16993/","number":"16993","commit":"809b32f9b2a8887cb39140f0b4db9271ceecd585"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17051/","number":"17051","commit":"231071a189a116afa963ca607110fdb161cbbddf"}]}
PIPELINE-->

jkelastic commented:
> I am so stoked to see this get fixed! Thanks @jkelastic.
> 
> Can we improve the UX by setting the default value for "min age" to 0? This requires changing three files: `store/defaults/cold_phase.js` (line 20), `store/defaults/warm_phase.js` (line 26), and `store/defaults/cold_phase.js` (line 18). All three lines need to be changed to:
> 
> ```js
> [PHASE_ROLLOVER_MINIMUM_AGE]: 0,
> ```
> 
> Now when you go into the form, all of these inputs should be prepopulated with 0 and you should be able to save the form.

@cjcenizal  No problem, I'm having a lot of fun and glad to be able to contribute to the EUI team. Third file you mentioned , do you mean `store/defaults/delete_phase.js` (line 18)?

I changed a few check conditions in the edit_policy.js to allow equal to and above 0. I also removed the warm phase test to check empty values as we've defaulted the input to be 0. Therefore, we'll always have a value and it wouldn't be empty.

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17065/)
* Commit: 3d23c4eab481464e0388301b67807fa28d0f7640

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17065/","number":"17065","commit":"3d23c4eab481464e0388301b67807fa28d0f7640"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17093/)
* Commit: 6393ae21df61785189ff4672de09e4b504b295c9

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17093/","number":"17093","commit":"6393ae21df61785189ff4672de09e4b504b295c9"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17339/)
* Commit: 595ab8f4e8da079cacb77ebd5f1d14d2a99b8590

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17339/","number":"17339","commit":"595ab8f4e8da079cacb77ebd5f1d14d2a99b8590"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17345/)
* Commit: 1827fab6c0f5bda33d6d705f0d22ded934267473

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17345/","number":"17345","commit":"1827fab6c0f5bda33d6d705f0d22ded934267473"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17349/)
* Commit: 2f7c197502d27f80ac4fcffd19af4410fcfc4276

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17349/","number":"17349","commit":"2f7c197502d27f80ac4fcffd19af4410fcfc4276"}]}
PIPELINE-->

jkelastic commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17353/)
* Commit: c0e0ba99f28aa694b0953491e233b8ae3b566904

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17353/","number":"17353","commit":"c0e0ba99f28aa694b0953491e233b8ae3b566904"}]}
PIPELINE-->

jkelastic commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/)
* Commit: 5765daa07a60e1b22c669a7e410e4a13aa9779e3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/","number":"17449","commit":"5765daa07a60e1b22c669a7e410e4a13aa9779e3"}]}
PIPELINE-->

jkelastic commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17621/)
* Commit: b76b7634aee232b8c634b2a85cacfdf1694ca643

### History
* :green_heart: [Build #17449](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/) succeeded 5765daa07a60e1b22c669a7e410e4a13aa9779e3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/","number":"17449","commit":"5765daa07a60e1b22c669a7e410e4a13aa9779e3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17621/","number":"17621","commit":"b76b7634aee232b8c634b2a85cacfdf1694ca643"}]}
PIPELINE-->

alisonelizabeth commented on x-pack/legacy/plugins/index_lifecycle_management/__jest__/components/edit_policy.test.js:
```diff
@@ -238,22 +237,14 @@ describe('edit policy', () => {
     });
   });
   describe('warm phase', () => {
-    test('should show number required error when trying to save empty warm phase', () => {
```
I think this is still a valid test. Is there a reason why you deleted it?

jkelastic replied:
I deleted it because we set a default value of 0 from the request below. Therefore, I don't think the test is valid anymore as will not have an empty value. 

> I am so stoked to see this get fixed! Thanks @jkelastic.
> 
> Can we improve the UX by setting the default value for "min age" to 0? This requires changing three files: `store/defaults/cold_phase.js` (line 20), `store/defaults/warm_phase.js` (line 26), and `store/defaults/cold_phase.js` (line 18). All three lines need to be changed to:
> 
> ```js
> [PHASE_ROLLOVER_MINIMUM_AGE]: 0,
> ```
> 
> Now when you go into the form, all of these inputs should be prepopulated with 0 and you should be able to save the form.

alisonelizabeth replied:
I believe the user could still delete the default value, in which case, the validation would still occur.

jkelastic replied:
thank you, updated my code

alisonelizabeth commented on x-pack/legacy/plugins/index_lifecycle_management/__jest__/components/edit_policy.test.js:
```diff
@@ -383,14 +374,14 @@ describe('edit policy', () => {
     });
   });
   describe('cold phase', () => {
-    test('should show positive number required error when trying to save cold phase with 0 for after', () => {
```
I think this is still a valid test as well. However, it would need to be updated to something like `'should show positive number required error when trying to save cold phase with negative number'`, and line 391 would need to be updated with a negative number.

jkelastic replied:
For all three `cold, warm, and delete` tests that checks for a positive number with 0 for after I've changed the test to       `expectedErrorMessages(rendered, [])` . I deleted the variable `positiveNumbersEqualAboveZeroErrorMessage` because now 0 is a valid input. Therefore, there wouldn't be any errors. When the jest ran it expected an error with `expectedErrorMessages(rendered, [positiveNumbersEqualAboveZeroErrorMessage])`

alisonelizabeth replied:
I understand. I was suggesting that testing for negative numbers is still valid.

alisonelizabeth replied:
Update: You can ignore this. I see we are already checking for negative numbers as a test below this:
```
    test('should show positive number required error when trying to save cold phase with -1 for after', () => {
```

alisonelizabeth commented on x-pack/legacy/plugins/index_lifecycle_management/__jest__/components/edit_policy.test.js:
```diff
@@ -464,14 +455,14 @@ describe('edit policy', () => {
     });
   });
   describe('delete phase', () => {
-    test('should show positive number required error when trying to save delete phase with 0 for after', () => {
```
Same comment as above. I believe this is still a valid test case, but should be updated to test with a negative number instead of `0`.

jkelastic replied:
Same response as above. There shouldn't be an error as `0` is now valid. I edited the test to check for no errors.

alisonelizabeth replied:
Update: You can ignore my original comment. I see there is a test to check for negative numbers below this test so I think we're all set:
```
    test('should show positive number required error when trying to save delete phase with -1 for after', () => {
```

alisonelizabeth commented on x-pack/legacy/plugins/index_lifecycle_management/public/store/selectors/lifecycle.js:
```diff
@@ -123,9 +130,9 @@ export const validatePhase = (type, phase, errors) => {
       } else if (
```
I was looking through the code again, and I'm wondering if this block of code is necessary. I don't think it will ever reach this point, as we're already checking for negative numbers on [L121](https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/index_lifecycle_management/public/store/selectors/lifecycle.js#L121).

```
} else if (phase[numberedAttribute] < 0) {
  phaseErrors[numberedAttribute] = [positiveNumberRequiredMessage];
}
```

Can you confirm?

jkelastic replied:
It will reach this point. Initially I thought the negative numbers on L121 would have caught it too, but it didn't.

alisonelizabeth replied:
Can you share how you were testing it to reach this?

jkelastic replied:
Sure, no problem. Initially I set <=0

```
} else if (phase[numberedAttribute] <= 0) {
  phaseErrors[numberedAttribute] = [positiveNumberRequiredMessage];
}
```

This will generate an error when I enter 0 through the Index Lifecycle Policy cold phase page. It was still not permitting 0. When I set < 0

```
} else if (phase[numberedAttribute] < 0) {
  phaseErrors[numberedAttribute] = [positiveNumberRequiredMessage];
}
```
I will not get an error when I enter 0 on the ILP page on the cold phase. Then when I run the jest test I will get a `received` error of 0 as 0 is now a valid value, but I will get an`expected` error of 1. 

```
expect(received).toBe(expected) // Object.is equality
    Expected: 1
    Received: 0
      74 | const expectedErrorMessages = (rendered, expectedErrorMessages) => {
      75 |   const errorMessages = rendered.find('.euiFormErrorText');
    > 76 |   expect(errorMessages.length).toBe(expectedErrorMessages.length);
         |                                ^
      77 |   expectedErrorMessages.forEach(expectedErrorMessage => {
      78 |     let foundErrorMessage;
      79 |     for (let i = 0; i < errorMessages.length; i++) {
```


Therefore, I edited the jest test case to allow no error to be returned by setting as the value 0 is permitted now

`expectedErrorMessages(rendered, [])`

alisonelizabeth replied:
Sorry, I wasn't very clear in my comment. I was referring to your statement:

> Initially I thought the negative numbers on L121 would have caught it too, but it didn't.

I was wondering what steps you were taking in the UI to reach this point. I am not able to reproduce.

jkelastic replied:
Sorry, my misunderstanding. Fixed the issue now.

jkelastic commented:
@elasticmachine merge upstream

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19088/)
* Commit: a2a5275914a223b59d26ee12a155d5255e7371ce

### History
* :green_heart: [Build #17621](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17621/) succeeded b76b7634aee232b8c634b2a85cacfdf1694ca643
* :green_heart: [Build #17449](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/) succeeded 5765daa07a60e1b22c669a7e410e4a13aa9779e3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/","number":"17449","commit":"5765daa07a60e1b22c669a7e410e4a13aa9779e3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17621/","number":"17621","commit":"b76b7634aee232b8c634b2a85cacfdf1694ca643"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19088/","number":"19088","commit":"a2a5275914a223b59d26ee12a155d5255e7371ce"}]}
PIPELINE-->

alisonelizabeth commented:
@elasticmachine merge upstream

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19269/)
* Commit: c382972a46c087f8128d0b17527d8d61f27027c0

### History
* :broken_heart: [Build #19088](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19088/) failed a2a5275914a223b59d26ee12a155d5255e7371ce
* :green_heart: [Build #17621](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17621/) succeeded b76b7634aee232b8c634b2a85cacfdf1694ca643
* :green_heart: [Build #17449](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/) succeeded 5765daa07a60e1b22c669a7e410e4a13aa9779e3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/","number":"17449","commit":"5765daa07a60e1b22c669a7e410e4a13aa9779e3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17621/","number":"17621","commit":"b76b7634aee232b8c634b2a85cacfdf1694ca643"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19088/","number":"19088","commit":"a2a5275914a223b59d26ee12a155d5255e7371ce"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19269/","number":"19269","commit":"c382972a46c087f8128d0b17527d8d61f27027c0"}]}
PIPELINE-->

alisonelizabeth commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19296/)
* Commit: 1f0638073780f51db1678dec0cf50b74e4f8fc14

### History
* :broken_heart: [Build #19269](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19269/) failed c382972a46c087f8128d0b17527d8d61f27027c0
* :broken_heart: [Build #19088](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19088/) failed a2a5275914a223b59d26ee12a155d5255e7371ce
* :green_heart: [Build #17621](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17621/) succeeded b76b7634aee232b8c634b2a85cacfdf1694ca643
* :green_heart: [Build #17449](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/) succeeded 5765daa07a60e1b22c669a7e410e4a13aa9779e3

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17449/","number":"17449","commit":"5765daa07a60e1b22c669a7e410e4a13aa9779e3"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/17621/","number":"17621","commit":"b76b7634aee232b8c634b2a85cacfdf1694ca643"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19088/","number":"19088","commit":"a2a5275914a223b59d26ee12a155d5255e7371ce"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19269/","number":"19269","commit":"c382972a46c087f8128d0b17527d8d61f27027c0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/19296/","number":"19296","commit":"1f0638073780f51db1678dec0cf50b74e4f8fc14"}]}
PIPELINE-->

alisonelizabeth commented:
@cjcenizal do you think you could take another look at @jkelastic's latest changes?

jkelastic commented:
> Great work @jkelastic! Thanks for doing this.

No problem! Thanks for giving me the opportunity! I had a great time, fun, and learned lots!
